### ISSUE TITLE
Running multiple servers.

### ISSUE THREAD

**Issue by KunalBurangi** [SAP/PyRFC#284]
Here's what i a trying to do. 
I am trying to create multiple servers and server them simultaniously

**To Reproduce**
code to reproduce 
`import os
from pyrfc import Server, set_ini_file_directory, Connection
import requests

# function which is added in server 
# this function gets called when Z_RJAL_SERVERTEST RFC is executed.
# this function then hit Integration service APi which saves the received data from RFC
# and then responds back with what ever responce was received. 
def my_custom_rfc1(request_context=None, DATA=[]):
    print("my_custom_rfc1 invoked ")
    # print("request_context", request_context)
    print("DATA:",DATA)
    headers= {"accept":"application/json","transactionid":'Test1',"Content-Type":"application/json"}
    url = 'http://localhost:3000/adapter/Inbound-Adapter-rfc-1'
    try:      
        print("Inside Try")   
        req =  requests.post(url,json = DATA, headers=headers)
        print(req.status_code, req.text)
    except:
        print("Inside except")   
        return {"STATUS":str(req.status_code),"TEXT":req.text}
    else:    
        print("Inside else")
        return {"STATUS":str(req.status_code),"TEXT":req.text}

# function which is added in server 
# this function gets called when Z_RJAL_SERVERTEST RFC is executed.
# this function then hit Integration service APi which saves the received data from RFC
# and then responds back with what ever responce was received. 
def my_custom_rfc2(request_context=None,DATA=[]):
    print("my_custom_rfc2 invoked ")
    # print("request_context", request_context)

    print("DATA:",DATA)
    headers= {"accept":"application/json","transactionid":'Test2',"Content-Type":"application/json"}
    url = 'http://localhost:3000/adapter/Inbound-Adapter-rfc-2'
    try:      
        print("Inside Try")   
        req =  requests.post(url,json = DATA, headers=headers)
        print(req.status_code, req.text)
    except:
        print("Inside except")   
        return {"STATUS":str(req.status_code),"TEXT":req.text}
    else:    
        print("Inside else")
        return {"STATUS":str(req.status_code),"TEXT":req.text}
# function which is added in server 
# this function gets called when Z_RJAL_SERVERTEST RFC is executed.
# this function then hit Integration service APi which saves the received data from RFC
# and then responds back with what ever responce was received. 
def my_custom_rfc3(request_context=None,DATA=[]):
    print("my_custom_rfc3 invoked ")
    # print("request_context", request_context)
    print("DATA:",DATA)
    headers= {"accept":"application/json","transactionid":"Test3","Content-Type":"application/json"}
    url = 'http://localhost:3000/adapter/Inbound-Adapter-rfc-3'
    try:      
        print("Inside Try")   
        req =  requests.post(url,json = DATA, headers=headers)
        print(req.status_code, req.text)
    except:
        print("Inside except")   
        return {"STATUS":str(req.status_code),"TEXT":req.text}
    else:    
        print("Inside else")
        return {"STATUS":str(req.status_code),"TEXT":req.text}
# Function that starts the server
# It also adds the function to the server which will get called fro RFC
# here the server starts and serve 
def main():
    dir_path = os.path.dirname(os.path.realpath(__file__))
    set_ini_file_directory(dir_path)
    
    server1 = Server({"dest": "gateway"}, {"dest": "client"},{"server_log":True})
    server1.add_function("Z_RJAL_SERVERTEST", my_custom_rfc1)
    
    server2 = Server({"dest": "gateway"}, {"dest": "client"},{"server_log":True})
    server2.add_function("Z_RJAL_SERVERTEST_1", my_custom_rfc2)
    
    server3 = Server({"dest": "gateway"}, {"dest": "client"},{"server_log":True})
    server3.add_function("Z_RJAL_SERVERTEST_2", my_custom_rfc3)
    
    print("\nPress CTRL-C to skip server test...")
    server1.serve(20)
    server2.serve(20)
    server3.serve(20)
    
main()`

`**Screenshots**`
![image](https://user-images.githubusercontent.com/37167732/196194374-ee5aa16e-c931-430a-a510-e61d38426d23.png)

**Environment**
 - OS: Windows 10
 - PyRFC version - latest one

**Additional context**
It seems like it only started one server and all the functions were added to it. Whereas code wise it looks like one server one function 
(According to my code)

How can I circumvent this problem? 

Is it even possible to run multiple servers? 

Also if not can I add multiple hosts and clients to one server? 
in sapnwrfc.ini?

**Comment by KunalBurangi** [comment]
@bsrdjan can you please help

**Comment by mpf82** [comment]
From what I can tell, the 3 functions are added to 3 different server instances (see "Server object at X", where X has 3 different values in your screenshot).

Also, Server.serve() starts a HTTPServer server with serve_forever(), so any other calls in the same thread (in this case your main thread) are halted, until the server shuts down, that's why in the log you only see a notification of your first server starting.

So much for a first analyses of what's happening.

I guess you could start each server instance in its own thread, but maybe there are better options.

**Comment by KunalBurangi** [comment]
UPDATE - 
what @mpf82  is saying is correct. 
but in that case, I was still able to execute RFC which was handled by server 2 and server 3 ... 
which means there is a single pool of functions and all the servers have access to all the functions. 
is that correct?

**Comment by bsrdjan** [comment]
Looking into example I see all handles are different. What exactly leads to conclusion:

> It seems like it only started one server and all the functions were added to it.

**Comment by KunalBurangi** [comment]
yes, but what I want is multiple servers that listen to different functions.
Is it possible...

**Comment by KunalBurangi** [comment]
so let's say I have 2 SAP system let's say SAP1 and SAP2 which wants to communicate to my application using RFC functions 
now I want to create 2 RFC servers say RFCServer1 and RFCServer2 (using pyrfc module ) RFCServer1 will communicate to SAP1 and RFCServer2  With SAP2.

what happens right now ... 

now both sap systems will have a different configuration which I have to configure in the sapnwrfc.ini file.
which I will create in runtime and provide it... 

what happens right now is the RFC server which starts last works... but the first doesn't work...  it seems like the first server was overridden by my 2nd
( the example I provided in this thread was just a sample code... not what I am mentioning here... )

**Comment by bsrdjan** [comment]
Yes, the scenario is supported by PyRFC and RFC SDK design. 

As @mpf82 mentioned, threads or multiple Python scripts can be used, each with one server instance. In either case all servers share the same SAP NWRFC SDK instance on host system. 

Here one example:

```python
import os
from pyrfc import Server, set_ini_file_directory
from threading import Thread

# server functions

def my_stfc_connection(request_context=None, REQUTEXT=""):
    print("stfc invoked")
    print("request_context", request_context)
    print(f"REQUTEXT: {REQUTEXT}")

    return {"ECHOTEXT": REQUTEXT, "RESPTEXT": "Python server here"}

def my_bapi_user(request_context=None, USERNAME=""):
    print("bapi user invoked")
    print("request_context", request_context)
    print(f"USERNAME: {USERNAME}")

    return {"USERNAME": USERNAME, "MESSAGE": "Python server here"}

# server authorisation check

def my_auth_check(func_name=False, request_context={}):
    print(f"authorization check for '{func_name}'")
    print("request_context", request_context)
    return 0

dir_path = os.path.dirname(os.path.realpath(__file__))
set_ini_file_directory(dir_path)

def server1_serve():
    # create server for ABAP system ABC
    server = Server({"dest": "gateway"}, {"dest": "ABC"}, {"port": 8081, "server_log": False})

    # expose python function my_stfc_connection as ABAP function STFC_CONNECTION, to be called by ABAP system
    server.add_function("STFC_CONNECTION", my_stfc_connection)

    # start server
    server.serve()


def server2_serve():
    # create server for ABAP system XYZ
    server = Server({"dest": "gateway"}, {"dest": "XYZ"}, {"port": 8081, "server_log": False})

    # expose python function my_bapi_user as ABAP function BAPI_USER_GET_DETAIL, to be called by ABAP system
    server.add_function("BAPI_USER_GET_DETAIL", my_bapi_user)

    # start server
    server.serve()

# start servers 

server1_thread = Thread(target=server1_serve)
server1_thread.start()
print("Server1 started")

server2_thread = Thread(target=server2_serve)
server2_thread.start()
print("Server2 started")
```

**Comment by Ulrich-Schmidt** [comment]
In addition to what bsrdjan said: note also that the underlying NW RFC Library manages server functions not by "server object", but by SAP System ID. So, if you create two servers against the same system ("ABC"), these two servers will share each other's server functions.

This explains, why the one server that you started in your original example (as others already explained: you startet only one server, because the server.serve() method blocks until the server is stopped again... So, server 2 and server 3 were never started in your code... :-) ), saw also the server functions of servers 2 and 3: all three servers were started against the same backend. So they share the registered function modules.

**Comment by KunalBurangi** [comment]
@Ulrich-Schmidt , @bsrdjan  ok,  Thanks you so much. one final question.. 
how are the ports used.. 
Here is another use case 
OS - Linux. 
so I tried doing the same as @bsrdjan  you mentioned.. and that works fine... 
If I don't provide the port then in Linux it gives me an error that the address is already in use. 
I checked the pyrfc code as well and found out that 8081 is the default port so when I try to start multiple servers it gives an error which is fine. 
but just wanted to know how are these ports getting exposed to the outer world.
I mean how is an SAP SYSTEM able to communicate with a PORT on my local system when it is not exposed to the outer world?

**Comment by Ulrich-Schmidt** [comment]
I only know the RFC Library part, not the Python part. And when you start a "registered" server using the RFC library, it does **not** open a port! (From TCP/IP point of view, a "registered server" is actually a client...)

But from what I learned from bsrdjan, it appears that the Python implementation of the Server object does two things at the same time:

1. It starts a registered RFC server on the backend, using libsapnwrfc.so (which opens a TCP/IP connection to the SAP system)
2. It starts an HTTP server, which then opens a listening port.

You can then connect to the HTTP server with a browser and shutdown the RFC server. But from RFC point of view, this HTTP server is not needed and not used at all. (Backend system does not know anything about this HTTP server...)

I don't know, why PyRFC made this design decision to start an HTTP server in addition to every RFC server. In my opinion it is a bit "strange" and over-complicated?! I would have expected the Server object to simply have two methods for start and stop, and then the RFC server can simply be stopped from Python code, instead of via an HTTP request?!

**Comment by bsrdjan** [comment]
> I don't know, why PyRFC made this design decision to start an HTTP server in addition to every RFC server. In my opinion it is a bit "strange" and over-complicated?! I would have expected the Server object to simply have two methods for start and stop, and then the RFC server can simply be stopped from Python code, instead of via an HTTP request?!
(Perhaps such a "shutdown via method call" is also possible, and the shutdown via HTTP request is only an "additional option" for administrators of the Python application?!)

http was requested by first user but indeed over-complicated approach for the majority. Shutdown and status check  shall be provided via method calls, by the end of this year hopefully

**Comment by KunalBurangi** [comment]
@bsrdjan 
Ok, Thanks, 
Also, can we just improve the documentation? 
the documentation does not have anything related to the server functionality.
It is very hard to understand.
I would be happy to contribute if you want.

**Comment by bsrdjan** [comment]
Hello @KunalBurangi, contributions are welcome :) You can add the server section to existing documentation, using [node-rfc server documentation](https://github.com/SAP/node-rfc/blob/main/doc/usage.md#server) as example. Building the Sphinx documentation is described [here](http://sap.github.io/PyRFC/build.html#building-the-documentation). Alternatively, the whole documentation could be ported to Markdown.

I removed the http server from PyRFC, tested with below script and pushed the change to main branch. 

Can you build from source and test with your systems? 

If build from source is a challenge in your environment, just let me know your platform and I will send the wheel.

```python
import os
from pyrfc import Server, set_ini_file_directory
from threading import Thread

# server functions


def my_stfc_connection(request_context=None, REQUTEXT=""):
    print("stfc invoked")
    print("request_context", request_context)
    print(f"REQUTEXT: {REQUTEXT}")

    return {"ECHOTEXT": REQUTEXT, "RESPTEXT": "Python server here"}


def my_bapi_user(request_context=None, USERNAME=""):
    print("bapi user invoked")
    print("request_context", request_context)
    print(f"USERNAME: {USERNAME}")

    return {"USERNAME": USERNAME, "MESSAGE": "Python server here"}


# server authorisation check
def my_auth_check(func_name=False, request_context={}):
    print(f"authorization check for '{func_name}'")
    print("request_context", request_context)
    return 0


dir_path = os.path.dirname(os.path.realpath(__file__))
set_ini_file_directory(dir_path)


# create server for ABAP system ABC
server1 = Server({"dest": "gateway"}, {"dest": "MME"}, {"port": 8081, "server_log": False})

# expose python function my_stfc_connection as ABAP function STFC_CONNECTION, to be called by ABAP system
server1.add_function("STFC_CONNECTION", my_stfc_connection)

# start server 1
server1.serve()

# create server for ABAP system XYZ
server2 = Server({"dest": "gateway"}, {"dest": "QM7"}, {"port": 8081, "server_log": False})

# expose python function my_bapi_user as ABAP function BAPI_USER_GET_DETAIL, to be called by ABAP system
server2.add_function("BAPI_USER_GET_DETAIL", my_bapi_user)

# start server 2
server2.serve()


# stop servers

input("Press Enter to stop server 1...")
server1.close()

input("Press Enter to stop server 2...")
server2.close()
```

**Comment by KunalBurangi** [comment]
@bsrdjan 
ok, I ll test it and let you know.

**Comment by KunalBurangi** [comment]
@bsrdjan ,
I will still have to run 2 servers in 2 different processes/ threads right? 
because I tested the code that you provided and it seems to be inconsistent.
sometimes both server works sometimes only one works. 
if I stop one both stop working... 

also, I am seeing these errors in terminal - i am not sure if this is the reason. 
![image](https://user-images.githubusercontent.com/37167732/198289000-fd10b874-dba7-4aa6-8269-377b97f59728.png).

i also tested it with sub process / threads but the subprocess closes itself . It is not staying alive. 
i am assuming that the serve process is returning and hence the process closes as the function executed completely.

i was able to test it running in subprocesses by doing some work arounds like while true or time.sleep() 
but still it was inconsistent. 
I am guessing that somehow the TCP/IP connection is breaking... I am not sure... 

`import os
from pyrfc import Server, set_ini_file_directory, Connection
import requests
import threading  
import time
from multiprocessing import Process, log_to_stderr, set_start_method
import threading

def my_custom_rfc1(request_context=None, DATA=[]):
    print("my_custom_rfc1 invoked ")
    # print("request_context", request_context)
    print("DATA:",DATA)

def my_custom_rfc2(request_context=None,DATA=[]):
    print("my_custom_rfc2 invoked ")
    # print("request_context", request_context)

    print("DATA:",DATA)

def my_custom_rfc3(request_context=None,DATA=[]):
    print("my_custom_rfc3 invoked ")
    print("DATA:",DATA)

def main():
    dir_path = os.path.dirname(os.path.realpath(__file__))
    set_ini_file_directory(dir_path)
    
  
    if __name__ == '__main__':
      try:
        t1 = Process( target = startNewServer,args = ("Z_RJAL_SERVERTEST",my_custom_rfc1,"gateway","client"))
        t1.start()
        t2 = Process( target  = startNewServer,args =("Z_RJAL_SERVERTEST_1",my_custom_rfc2,"gateway1","client1"))
        t2.start()
      except Exception as e:
        print("Error orroured")  
        print(e)
    
    

def startNewServer(rfcName,callbackfunctionName,gateway,client):
    server = Server({"dest": gateway}, {"dest":client},{"server_log":True})
    server.add_function(rfcName,callbackfunctionName)
    server.serve()
    time.sleep(600)
    print("\nPress CTRL-C to skip server test...")

`

**Comment by bsrdjan** [comment]
My previous example was illustrative (not really tested) and here the one tested with two ABAP systems calling the same Python server:

```python
import os
from pyrfc import Server, set_ini_file_directory
from threading import Thread

# server functions


def my_stfc_connection(request_context=None, REQUTEXT=""):
    print("stfc connection invoked")
    print("request_context", request_context)
    print(f"REQUTEXT: {REQUTEXT}")
    return {"ECHOTEXT": REQUTEXT, "RESPTEXT": "Python server 1 here"}


def my_stfc_structure(request_context=None, IMPORTSTRUCT={}, RFCTABLE=[]):
    print("stfc structure invoked")
    print("request_context", request_context)
    print(f"IMPORTSTRUCT: {IMPORTSTRUCT}")
    return {"ECHOSTRUCT": {}, "RFCTABLE": [], "RESPTEXT": "Python server 2 here"}


# server authorisation check


def my_auth_check(func_name=False, request_context={}):
    print(f"authorization check for '{func_name}'")
    print("request_context", request_context)
    return 0


dir_path = os.path.dirname(os.path.realpath(__file__))
set_ini_file_directory(dir_path)

# create server for ABAP system MME
server1 = Server({"dest": "gateway"}, {"dest": "MME"}, {"port": 8081, "server_log": False})
# expose python function my_stfc_connection as ABAP function STFC_CONNECTION, to be called by ABAP system
server1.add_function("STFC_CONNECTION", my_stfc_connection)
# start server 1
server1.serve()
print("server1 started")

# create server for ABAP system XYZ
server2 = Server({"dest": "gatewayqm7"}, {"dest": "QM7"}, {"port": 8081, "server_log": False})
# expose python function my_bapi_user as ABAP function BAPI_USER_GET_DETAIL, to be called by ABAP system
server2.add_function("STFC_STRUCTURE", my_stfc_structure)
# start server 2
server2.serve()
print("server2 started")

# serve requests
input("Press Enter to stop servers ...")
server1.close()
server2.close()
```

I tested with your script and it works fine as well:

```python
import os
from pyrfc import Server, set_ini_file_directory, Connection
import requests
import threading
import time
from multiprocessing import Process, log_to_stderr, set_start_method
import threading


def my_stfc_connection(request_context=None, REQUTEXT=""):
    print("stfc connection invoked")
    print("request_context", request_context)
    print(f"REQUTEXT: {REQUTEXT}")
    return {"ECHOTEXT": REQUTEXT, "RESPTEXT": "Python server 1 here"}


def my_stfc_structure(request_context=None, IMPORTSTRUCT={}, RFCTABLE=[]):
    print("stfc structure invoked")
    print("request_context", request_context)
    print(f"IMPORTSTRUCT: {IMPORTSTRUCT}")
    return {"ECHOSTRUCT": {}, "RFCTABLE": [], "RESPTEXT": "Python server 2 here"}


def main():
    dir_path = os.path.dirname(os.path.realpath(__file__))
    set_ini_file_directory(dir_path)

    if __name__ == "__main__":
        try:
            t1 = Process(target=startNewServer, args=("STFC_CONNECTION", my_stfc_connection, "gateway", "MME"))
            t1.start()
            t2 = Process(target=startNewServer, args=("STFC_STRUCTURE", my_stfc_structure, "gatewayqm7", "QM7"))
            t2.start()
        except Exception as e:
            print("Error orroured")
            print(e)


def startNewServer(rfcName, callbackfunctionName, gateway, client):
    server = Server({"dest": gateway}, {"dest": client}, {"server_log": True})
    server.add_function(rfcName, callbackfunctionName)
    server.serve()
    time.sleep(600)
    print("\nPress CTRL-C to skip server test...")

main()
```

<img width="599" alt="Screenshot 2022-10-28 at 13 11 12" src="https://user-images.githubusercontent.com/1697847/198573885-83712cf5-6c46-4dc4-8acb-0f05a2dece23.png">

To verify your Python server logic and configuration, you can first test with simple SAP standard RFC modules, like STFC* above. It proofs the "infrastructure" is fine, like network and servers' configurations.

When that works, you try with more complex and custom RFCs. The `iteritem` error indicates something with input/output parameters is wrong. In ABAP there are mandatory and optional parameters and your "my_custom*" functions return nothing, which is a bit strange because. You can check the minimum of parameters required on ABAP side and add them as input and/or output to Python functions, then it should work. You can take STFC_CONNECTION/STRUCTURE as example.

**Comment by bsrdjan** [comment]
> I will still have to run 2 servers in 2 different processes/ threads right?

I am not sure about it, could not find the information in [SAP NetWeaver RFC SDK ProgrammingGuide](https://support.sap.com/content/dam/support/en_us/library/ssp/products/connectors/nwrfcsdk/NW_RFC_750_ProgrammingGuide.pdf) on  https://support.sap.com/en/product/connectors/nwrfcsdk.html and in [NW RFC SDK Part 2: RFC Server Programs](http://go.sap.com/docs/download/2016/09/1c9a1fe1-8a7c-0010-82c7-eda71af511fa.pdf) PDF article listed here: https://wiki.scn.sap.com/wiki/display/ABAPConn/SAP+NetWeaver+RFC+SDK+--+RFC+Server+Programs

@Ulrich-Schmidt, can you help us clarify if RFC SDK server spawns own threads or application shall handle? In [NW RFC SDK Part 2: RFC Server Programs](http://go.sap.com/docs/download/2016/09/1c9a1fe1-8a7c-0010-82c7-eda71af511fa.pdf) the "blocking wait" is mentioned. I tested two RFC SDK servers in the same thread in Python (example above) and did not experience blocking when calling them from two ABAP systems. Therefore not sure :)

**Comment by Ulrich-Schmidt** [comment]
Depends on which RFC API the Python implementation uses internally.

- If you use the "old way" with RfcListenAndDispatch() API, you have to spawn your own thread for every server connection and run a "dispatch loop" in that thread.
- If you use the "new way" with RfcLaunchServer() API (available since 2018, release 7.50), then libsapnwrfc.so starts its own thread pool internally, which is then used for dispatching the incoming RFC calls. In that case you can start (and stop) multiple servers from one single Python thread. (And thread management for the different servers and all their connections is handled automatically by libsapnwrfc.so on C/C++ level.)

The PDF article you cite above, is from 2007, so it only describes the old "listen&dispatch approach". For the new "launch approach", you need to follow the ProgrammingGuide (chapter 5.2.2). I would definitely recommend using the new approach, because then the code on Python side becomes really easy, as the RFC lib is already doing most of the work. (Basically, it is already doing everything automatically, which with the old approach you would have to do by yourself:
- as already mentioned, the thread management, 
- the error handling
- the proper "re-connect behavior" in case connections break down due to system failures or network problems.)

Best Regards, Ulrich

**Comment by bsrdjan** [comment]
PyRFC is using "new way"',  it explains why not blocking

https://github.com/SAP/PyRFC/blob/2dacc4a6f1ce8342fd3381df2a15e866527a9118/src/pyrfc/_pyrfc.pyx#L1520

**Comment by KunalBurangi** [comment]
@bsrdjan,
I have tested it several times with STFC_CONNECTION and STFC_STRUCTURE.
it does work..... but not all the time... 
even I only tried with one server and one function. 
still, it doesn't work all the time.
3 out of 5 times it worked but 2  times it failed. the behavior is entirely random.
![image](https://user-images.githubusercontent.com/37167732/198605633-3cb577e9-90a9-4747-9ed6-c6399440582e.png)
can you try testing it again and again in a short time duration?

**Comment by bsrdjan** [comment]
How do you test from ABAP? Can you share ABAP test report with STFC modules, that I can replicate ?

**Comment by Ulrich-Schmidt** [comment]
The screen looks like SE37.

But a report would indeed be better, because SE37 does not display the error details that come with this SYSTEM_FAILURE. We would need to see these details. If you activate RFC trace (best with a high level like 3), we can see the error details in the trace.

**Comment by KunalBurangi** [comment]
I don't have any experience in SAP  so I have very little knowledge about it and I am not sure what you mean by the ABAP test report.  do you mean the source code? 
this is it 
![image](https://user-images.githubusercontent.com/37167732/198635584-4d7225ec-33a9-4f3a-af5a-c783aad9dc7b.png)
I am just executing this code.
The destination I have used is "NWRFC" which I have created in SM59. 
yes the RFC function is inside SE37. 
I have activated the RFC trace from the destination 
but I am not sure where to check the traces. 
in the python terminal, I can see this when the whole thing works. but nothing when we get an error.
![image](https://user-images.githubusercontent.com/37167732/198647589-5c45be2e-c628-4363-a05f-892775401583.png)

for proper trace, I would have to take some help from my colleague. who is familiar with SAP which might happen on Monday.

**Comment by Ulrich-Schmidt** [comment]
PS: you can see the difference between the ABAP code and the Python code, if you just clear the field "RFC target sys" in SE37. If that field is blank, the FM is executed locally, and you see the output of the ABAP code, while if that field is set to "NWRFC", the Python code gets executed, and you will see the output of that Python code in SE37.

**Comment by KunalBurangi** [comment]
> > I am just executing this code.
> 
> No, you are not... You are executing Python code! (The ABAP code that your screenshot shows, is executed only, when you execute that function module inside an SAP system, but you are not doing that, you are executing that function module on the destination "NWRFC", which points to your Python program! So in that case the Python code of "my_stfc_connection()" gets executed, not the ABAP code... :-) )
> 
> The trace file can be found in the current working directory of your Python program. Look for two files: dev_rfc.log and rfc_xxxx.trc

yes I mean I am executing that function from the screenshot. 
the dev_rfc.log does not get any new entry after executing the function 
and here  is the trace file. 


```

**** Trace file opened at 2022-10-28 19:58:25.764968 UTC+05:30 (India Standard Time), Encoding UTF-8
NW RFC Library: SDK variant, Release 750 Patch Level 10, Compiled on Jun 11 2022 06:51:55
 CPIC library: 753.2021.05.10 version 3, NI library: 40, Kernel Release: 753 Patch Level 1021
 Current working directory: C:\Private Projects\rfc-python-sample, Program: python
 Hardware: PC with Windows NT 8x AMD64 Level 6 (Mod 142 Step 12), Operating_system: Windows NT 10.0
 Hostname: PC21IE, IP address: 192.168.0.100, IPv6 address: 2a02:24d8:77:712::c4

2022-10-28 19:58:25.765770 [42324] 	>> RfcConnection::readUpTo
Received RFCID 0x0101 (257), 8 bytes
Received RFCID 0x0103 (259), 4 bytes
Received RFCID 0x0106 (262), 11 bytes
Received RFCID 0x0007 (7), 15 bytes
Received RFCID 0x0018 (24), 10 bytes
Received RFCID 0x0008 (8), 13 bytes
Received RFCID 0x0011 (17), 1 byte
Received RFCID 0x0013 (19), 4 bytes
Received RFCID 0x0012 (18), 4 bytes
Received RFCID 0x0006 (6), 5 bytes
Received RFCID 0x0130 (304), 32 bytes
Received RFCID 0x0111 (273), 4 bytes
Received RFCID 0x0114 (276), 3 bytes
Received RFCID 0x0115 (277), 1 byte
Received RFCID 0x0009 (9), 4 bytes
Skipping RFCID 0x0009 (9), 4 bytes
Received RFCID 0x0134 (308), 3 bytes
Received RFCID 0x0501 (1281), 1 byte
Skipping RFCID 0x0501 (1281), 1 byte
Received RFCID 0x0136 (310), 37 bytes
Skipping RFCID 0x0136 (310), 37 bytes
Received RFCID 0x0502 (1282), 0 bytes
Skipping RFCID 0x0502 (1282), 0 bytes
Received RFCID 0x000B (11), 3 bytes
Skipping RFCID 0x000B (11), 3 bytes
Received RFCID 0x0102 (258), 15 bytes
2022-10-28 19:58:25.777100 [42324] 	<< RfcConnection::readUpTo returned 0

Dispatching function STFC_CONNECTION on handle 2049819846096
Looking up ServerFunction for STFC_CONNECTION from ECH.
Looking up ServerFunction for STFC_CONNECTION from ECH.
Found user function in cache for ECH.
2022-10-28 19:58:25.779089 [42324] 	>> RfcFunction::rfcDeserialize
2022-10-28 19:58:25.779274 [42324] 	>> RfcConnection::readUpTo
Received RFCID 0x0337 (823), 0 bytes
Skipping RFCID 0x0337 (823), 0 bytes
Received RFCID 0x0503 (1283), 0 bytes
Skipping RFCID 0x0503 (1283), 0 bytes
Received RFCID 0x0131 (305), 230 bytes
Received RFCID 0x0514 (1300), 16 bytes
Skipping RFCID 0x0514 (1300), 16 bytes
Received RFCID 0x0420 (1056), 4 bytes
Skipping RFCID 0x0420 (1056), 4 bytes
Received RFCID 0x0512 (1298), 0 bytes
Skipping RFCID 0x0512 (1298), 0 bytes
Received RFCID 0x0205 (517), 8 bytes
Received RFCID 0x0205 (517), 8 bytes
Received RFCID 0x0201 (513), 8 bytes
Received RFCID 0x0203 (515), 255 bytes
Received RFCID 0x0104 (260), 313 bytes
Skipping RFCID 0x0104 (260), 313 bytes
Received RFCID 0xFFFF (65535), 0 bytes
Skipping RFCID 0xFFFF (65535), 0 bytes
2022-10-28 19:58:25.781720 [42324] 	<< RfcConnection::readUpTo returned 0

2022-10-28 19:58:25.781813 [42324] 	<< RfcFunction::rfcDeserialize returned 0

>> RfcFunction name="STFC_CONNECTION"
>> IMPORTING parameters
>> REQUTEXT, RFCTYPE_CHAR, Length: 510, Active: true, Requested: true
	Value: TEST1
<< REQUTEXT

<< IMPORTING parameters

>> EXPORTING parameters
>> ECHOTEXT, RFCTYPE_CHAR, Length: 510, Active: true, Requested: true
	Value: 
<< ECHOTEXT

>> RESPTEXT, RFCTYPE_CHAR, Length: 510, Active: true, Requested: true
	Value: 
<< RESPTEXT

<< EXPORTING parameters

>> CHANGING parameters
<< CHANGING parameters

>> TABLES parameters
<< TABLES parameters

<< RfcFunction

2022-10-28 19:58:25.783825 [42324] >> Invoking server function implementation (user function)
2022-10-28 19:58:25.783971 [42324] >> RfcDescribeFunction
RfcDescribeFunction(000000216818CE30, 0000000000000000)
RfcDescribeFunction(000000216818CE30) returns 000001DD42BAE200: STFC_CONNECTION (origin=ECH)
2022-10-28 19:58:25.784899 [42324] << RfcDescribeFunction returned RFC_OK

2022-10-28 19:58:25.785061 [42324] >> RfcGetConnectionAttributes
2022-10-28 19:58:25.785178 [42324] << RfcGetConnectionAttributes returned RFC_OK

2022-10-28 19:58:25.790449 [42324] << Invoking server function implementation (user function) returned RFC_OK

>> HEXDUMP
2022-10-28 19:58:25.790938 [42324] Sending data for handle 2049819846096, convID 70075706, codepage 4103
000000 | 01010008 03010105 04010003 01010103 |................|
000010 | 00040000 0E1B0103 0106000B 04010003 |................|
000020 | 000A0200 00002301 06000700 1A310039 |......#......1.9|
000030 | 0032002E 00310036 0038002E 0030002E |.2...1.6.8...0..|
000040 | 00310030 00300000 07001800 28320061 |.1.0.0......(2.a|
000050 | 00300032 003A0032 00340064 0038003A |.0.2.:.2.4.d.8.:|
000060 | 00370037 003A0037 00310032 003A003A |.7.7.:.7.1.2.:.:|
000070 | 00630034 00001800 11000245 00001100 |.c.4.......E....|
000080 | 12000637 00350033 00001200 13001C37 |...7.5.3.......7|
000090 | 00350033 00200031 0030002E 00320038 |.5.3. .1.0...2.8|
0000a0 | 002E0030 002E0032 00310000 13013000 |...0...2.1....0.|
0000b0 | 0C700079 00740068 006F006E 00013002 |.p.y.t.h.o.n..0.|
0000c0 | 01001045 00430048 004F0054 00450058 |...E.C.H.O.T.E.X|
0000d0 | 00540002 01020301 FE540045 00530054 |.T......þT.E.S.T|
0000e0 | 00310020 00200020 00200020 00200020 |.1. . . . . . . |
0000f0 | 00200020 00200020 00200020 00200020 |. . . . . . . . |
000100 | 00200020 00200020 00200020 00200020 |. . . . . . . . |
000110 | 00200020 00200020 00200020 00200020 |. . . . . . . . |
000120 | 00200020 00200020 00200020 00200020 |. . . . . . . . |
000130 | 00200020 00200020 00200020 00200020 |. . . . . . . . |
000140 | 00200020 00200020 00200020 00200020 |. . . . . . . . |
000150 | 00200020 00200020 00200020 00200020 |. . . . . . . . |
000160 | 00200020 00200020 00200020 00200020 |. . . . . . . . |
000170 | 00200020 00200020 00200020 00200020 |. . . . . . . . |
000180 | 00200020 00200020 00200020 00200020 |. . . . . . . . |
000190 | 00200020 00200020 00200020 00200020 |. . . . . . . . |
0001a0 | 00200020 00200020 00200020 00200020 |. . . . . . . . |
0001b0 | 00200020 00200020 00200020 00200020 |. . . . . . . . |
0001c0 | 00200020 00200020 00200020 00200020 |. . . . . . . . |
0001d0 | 00200020 00200020 00200020 00200020 |. . . . . . . . |
0001e0 | 00200020 00200020 00200020 00200020 |. . . . . . . . |
0001f0 | 00200020 00200020 00200020 00200020 |. . . . . . . . |
000200 | 00200020 00200020 00200020 00200020 |. . . . . . . . |
000210 | 00200020 00200020 00200020 00200020 |. . . . . . . . |
000220 | 00200020 00200020 00200020 00200020 |. . . . . . . . |
000230 | 00200020 00200020 00200020 00200020 |. . . . . . . . |
000240 | 00200020 00200020 00200020 00200020 |. . . . . . . . |
000250 | 00200020 00200020 00200020 00200020 |. . . . . . . . |
000260 | 00200020 00200020 00200020 00200020 |. . . . . . . . |
000270 | 00200020 00200020 00200020 00200020 |. . . . . . . . |
000280 | 00200020 00200020 00200020 00200020 |. . . . . . . . |
000290 | 00200020 00200020 00200020 00200020 |. . . . . . . . |
0002a0 | 00200020 00200020 00200020 00200020 |. . . . . . . . |
0002b0 | 00200020 00200020 00200020 00200020 |. . . . . . . . |
0002c0 | 00200020 00200020 00200020 00200020 |. . . . . . . . |
0002d0 | 00200020 00200002 03020100 10520045 |. . . .......R.E|
0002e0 | 00530050 00540045 00580054 00020102 |.S.P.T.E.X.T....|
0002f0 | 0301FE50 00790074 0068006F 006E0020 |..þP.y.t.h.o.n. |
000300 | 00730065 00720076 00650072 00200031 |.s.e.r.v.e.r. .1|
000310 | 00200068 00650072 00650020 00200020 |. .h.e.r.e. . . |
000320 | 00200020 00200020 00200020 00200020 |. . . . . . . . |
000330 | 00200020 00200020 00200020 00200020 |. . . . . . . . |
000340 | 00200020 00200020 00200020 00200020 |. . . . . . . . |
000350 | 00200020 00200020 00200020 00200020 |. . . . . . . . |
000360 | 00200020 00200020 00200020 00200020 |. . . . . . . . |
000370 | 00200020 00200020 00200020 00200020 |. . . . . . . . |
000380 | 00200020 00200020 00200020 00200020 |. . . . . . . . |
000390 | 00200020 00200020 00200020 00200020 |. . . . . . . . |
0003a0 | 00200020 00200020 00200020 00200020 |. . . . . . . . |
0003b0 | 00200020 00200020 00200020 00200020 |. . . . . . . . |
0003c0 | 00200020 00200020 00200020 00200020 |. . . . . . . . |
0003d0 | 00200020 00200020 00200020 00200020 |. . . . . . . . |
0003e0 | 00200020 00200020 00200020 00200020 |. . . . . . . . |
0003f0 | 00200020 00200020 00200020 00200020 |. . . . . . . . |
000400 | 00200020 00200020 00200020 00200020 |. . . . . . . . |
000410 | 00200020 00200020 00200020 00200020 |. . . . . . . . |
000420 | 00200020 00200020 00200020 00200020 |. . . . . . . . |
000430 | 00200020 00200020 00200020 00200020 |. . . . . . . . |
000440 | 00200020 00200020 00200020 00200020 |. . . . . . . . |
000450 | 00200020 00200020 00200020 00200020 |. . . . . . . . |
000460 | 00200020 00200020 00200020 00200020 |. . . . . . . . |
000470 | 00200020 00200020 00200020 00200020 |. . . . . . . . |
000480 | 00200020 00200020 00200020 00200020 |. . . . . . . . |
000490 | 00200020 00200020 00200020 00200020 |. . . . . . . . |
0004a0 | 00200020 00200020 00200020 00200020 |. . . . . . . . |
0004b0 | 00200020 00200020 00200020 00200020 |. . . . . . . . |
0004c0 | 00200020 00200020 00200020 00200020 |. . . . . . . . |
0004d0 | 00200020 00200020 00200020 00200020 |. . . . . . . . |
0004e0 | 00200020 00200020 00200020 00200020 |. . . . . . . . |
0004f0 | 00020306 67000800 00000080 9EDD4006 |....g........Ý@.|
000500 | 67FFFF00 00FFFF00 00000000 00000000 |gÿÿ..ÿÿ.........|
<< HEXDUMP

>> RfcFunction name="STFC_CONNECTION"
>> IMPORTING parameters
>> REQUTEXT, RFCTYPE_CHAR, Length: 510, Active: true, Requested: true
	Value: TEST1
<< REQUTEXT

<< IMPORTING parameters

>> EXPORTING parameters
>> ECHOTEXT, RFCTYPE_CHAR, Length: 510, Active: true, Requested: true
	Value: TEST1
<< ECHOTEXT

>> RESPTEXT, RFCTYPE_CHAR, Length: 510, Active: true, Requested: true
	Value: Python server 1 here
<< RESPTEXT

<< EXPORTING parameters

>> CHANGING parameters
<< CHANGING parameters

>> TABLES parameters
<< TABLES parameters

<< RfcFunction


```

**Comment by Ulrich-Schmidt** [comment]
That is the trace of a successful call. If there is nothing else in the trace, when we have an unsuccessful call, that means that in that case the SYSTEM_FAILURE already happens on R/3 side, before the call reaches the Python server!
So in order to trouble shoot, you would have to check the error logs of the R/3 system. E.g. try ST22, this would show the details, if a dump is happening on SAP side, before the call reaches the Python side.
Or you could activate RFC trace on R/3 side (in the destination in SM59), and check whether that shows more details why the call is failing.

**Comment by KunalBurangi** [comment]
ok, Thanks for all the information. 
yes, it is not creating a trace file when it gives SYSTEM_FAILURE. 
I will try to check on the SAP side. 
is it possible that we are losing the connection frequently and then as you suggested it has re-connect behavior 
so it connects again and it works again?

**Comment by KunalBurangi** [comment]
I checked st22 but there we only see runtime errors and I am not able to see this error there..

**Comment by KunalBurangi** [comment]
@bsrdjan , 
did you check executing the RFC function multiple times?
also, 
I created the testing build using this article this is correct, right? 
https://holypython.com/python-packaging-local-installation-tests-before-uploading/ 

if not can you give me the build? as you were mentioning here 

> If build from source is a challenge in your environment, just let me know your platform and I will send the wheel.

I am testing it in Linux inside a docker container

**Comment by bsrdjan** [comment]
I tested with two ABAP servers, calling Python server in parallel and I could not reproduce the issue.

ABAP test script on both servers is the same, one is calling `STFC_CONNECTION` and another one `STFC_STRUCTURE`. Can you test with this or similar report?

```abap
*&---------------------------------------------------------------------*
*& Report ZSERVERTEST
*&---------------------------------------------------------------------*
*&
*&---------------------------------------------------------------------*
REPORT ZSERVERTEST.

DATA lv_struct LIKE  RFCTEST.
DATA lv_resp LIKE sy-lisel.
DATA lv_table LIKE TABLE OF RFCTEST.

DO 10 TIMES.
CALL FUNCTION 'STFC_STRUCTURE'  DESTINATION 'NODEJS'
  EXPORTING
    IMPORTSTRUCT       = lv_struct
 IMPORTING
   ECHOSTRUCT         = lv_struct
   RESPTEXT           = lv_resp
  TABLES
    RFCTABLE           = lv_table
     .

IF sy-subrc = 0.
  WRITE lv_resp.
ELSE.
  WRITE 'Server not reached'.
ENDIF.

ENDDO.
```

Linux with docker container is fine for testing and Linux wheels you can find here: https://github.com/SAP/PyRFC/releases/tag/v2.5.1

PyRFC Linux wheels are built on this [Ubuntu 18.04 docker container](https://github.com/SAP/fundamental-tools/blob/main/docker/ubuntu-qa.Dockerfile) and should work on other Linux distributions but build from source is official Python recommendation here. PyRFC build from source is described here https://github.com/SAP/PyRFC#download-and-installation

Before building from source, cython and wheels Python packages shall be installed and then you can clone the repository and build the wheel.

**Comment by KunalBurangi** [comment]
Hello @bsrdjan,
it works fine... 
The way I was testing earlier was the issue it seems ... 
I was using this method to create a local build [Link](https://holypython.com/python-packaging-local-installation-tests-before-uploading/)
which might be the problem.

now I have created a wheel as mentioned [here](https://github.com/SAP/PyRFC#download-and-installation) tested it again and this time it worked as expected.

**Comment by KunalBurangi** [comment]
@bsrdjan 
Also, can you create a task and assign it to me so I can contribute in providing the documentation

**Comment by Ulrich-Schmidt** [comment]
@bsrdjan: Could you add an EXCEPTIONS clause to the CALL FUNCTION statement in you little sample report? Because the way it currently is, it would crash with a shortdump, if the Python server is not reachable, instead of going into the ELSE block... :-)

You would need something like the following to catch the exceptions and assign SY-SUBRC:

`DATA lv_error_message(128).
...
   EXCEPTIONS
      COMMUNICATION_FAILURE = 1 MESSAGE lv_error_message
      SYSTEM_FAILURE = 2 MESSAGE lv_error_message`

BTW: how did you color some words red in your sample code?!?

**Comment by bsrdjan** [comment]
Good suggestion Ulrich, will add to example.

> BTW: how did you color some words red in your sample code?!?

I tried the keyword `abap` in GitHub syntax highlighting and it works, not sure since when.  Just use \`\`\`abap instead of \`\`\` at the code-block beginning.

**Comment by bsrdjan** [comment]
updated examples

```abap
report zservertest.

data lv_struct like  rfctest.
data lv_resp like sy-lisel.
data lv_table like table of rfctest.
data lv_error_message type char512.

do 10 times.
  call function 'STFC_STRUCTURE' destination 'NODEJS'
    exporting
      importstruct          = lv_struct
    importing
      echostruct            = lv_struct
      resptext              = lv_resp
    tables
      rfctable              = lv_table
    exceptions
      communication_failure = 1 message lv_error_message
      system_failure        = 2 message lv_error_message.

  if sy-subrc = 0.
    write lv_resp.
  else.
    write lv_error_message.
    exit.
  endif.

enddo.
```

and

```abap
report zservertest.

data lv_echo like sy-lisel.
data lv_resp like sy-lisel.
data lv_error_message type char512.

do 10 times.

  call function 'STFC_CONNECTION' destination 'NODEJS'
    exporting
      requtext              = 'XYZ'
    importing
      echotext              = lv_echo
      resptext              = lv_resp
    exceptions
      communication_failure = 1 message lv_error_message
      system_failure        = 2 message lv_error_message.

  if sy-subrc = 0.
    write lv_echo.
    write lv_resp.
  else.
    write lv_error_message.
    exit.
  endif.
enddo.
```

**Comment by KunalBurangi** [comment]
@bsrdjan 
Any Update on this thread?

**Comment by bsrdjan** [comment]
Hello @KunalBurangi,

to contribute to documentation, please refer to my comment https://github.com/SAP/PyRFC/issues/284#issuecomment-1292117044

After signing the contribution agreement you can add the server documentation and create the pull request.

**Comment by bsrdjan** [comment]
Two more server methods added: `start()` and `stop()`, to start/stop server in new thread. Example: [examples/threading/server.py](https://github.com/SAP/PyRFC/blob/main/examples/threading/server.py)

Already existing methods, `serve()` and `close()` work without threading and threading, http-server or similar control can be provided by application. Example: [examples/threading/server_own_threads.py](https://github.com/SAP/PyRFC/blob/main/examples/threading/server_own_threads.py)

Fixed by 68c0fe8499726fea3ebe7fe025a4b9e23fbfde0b

**Comment by KunalBurangi** [comment]
@bsrdjan , 
Thanks for all the support. 
just one question, today I got this error 
![image](https://user-images.githubusercontent.com/37167732/203474615-383cea30-3629-49fd-a022-fa7718bd9973.png)
Is there any limit on the number of TCP transactions on pyrfc side or it will be on SAP system side?

**Comment by bsrdjan** [comment]
It is ABAP server configuration https://stackoverflow.com/questions/59178676/max-no-of-200-conversations-exceeded-error-in-pyrfc

**Comment by KunalBurangi** [comment]
@bsrdjan okok Thanks, 
also as this thread is now closed are the changes released?

**Comment by bsrdjan** [comment]
the change will be included in next release. in the meantime can be tested by local build from source

**Comment by Ulrich-Schmidt** [comment]
Well, actually there is a limit on both sides, external client and ABAP server. In your case, you are hitting the client side limit:

"LOCATION: CPIC (TCP/IP) on local host ..."

I think, the ABAP side has a higher limit, like 500. But it can happen that you are not able to open even one connection, if other external programs or other SAP systems have already consumed the 500 available connections. I will update the stackoverflow question with the details.

BTW: is that really your hostname? It looks more like a pointer value is printed to the error message, instead of the content to which the pointer points.... May need a fix in the CPIC lib.

**Comment by KunalBurangi** [comment]
> Well, actually there is a limit on both sides, external client and ABAP server. In your case, you are hitting the client side limit:
> 
> "LOCATION: CPIC (TCP/IP) on local host ..."
> 
> I think, the ABAP side has a higher limit, like 500. But it can happen that you are not able to open even one connection, if other external programs or other SAP systems have already consumed the 500 available connections. I will update the stackoverflow question with the details.
> 
> BTW: is that really your hostname? It looks more like a pointer value is printed to the error message, instead of the content to which the pointer points.... May need a fix in the CPIC lib.

@Ulrich-Schmidt , 
so I would have to update it at both ends right? 
ABAP server and external client
I found this env variable 'CPIC_MAX_CONV ' this should work for external client right ?

**Comment by KunalBurangi** [comment]
@bsrdjan and @Ulrich-Schmidt ,
also why is it creating so many connections? 
I am just creating one server and shouldn't it have just one connection?

**Comment by Ulrich-Schmidt** [comment]
Yes, it should only have one connection (or as many as you specify in the REG_COUNT parameter).

BTW: I updated stackoverflow with details about maximum number of conversations.

**Comment by KunalBurangi** [comment]
> Yes, it should only have one connection (or as many as you specify in the REG_COUNT parameter).
> 
> BTW: I updated stackoverflow with details about maximum number of conversations.

oh ok . @Ulrich-Schmidt 
can you point me to the doc which explains all these parameters

**Comment by bsrdjan** [comment]
> can you point me to the doc which explains all these parameters

The doxygen documentation is here: https://support.sap.com/en/product/connectors/nwrfcsdk.html

@Ulrich-Schmidt, I could not find `RfcSetMaximumCpicConversations` in current doxygen documentation and also not in  `sapnwrfc.h`. Is public API?

**Comment by KunalBurangi** [comment]
Hi, 
@bsrdjan @Ulrich-Schmidt ,
Can we use SAP Outbound Queues with pyrfc ?

**Comment by Ulrich-Schmidt** [comment]
Hi,
you will need to use RfcInstallBgRfcHandlers() to install 5 callback functions that handle the different status change events and update the local transaction database. (Basics are explained in the sapnwrfc.h header file. A detailed description with code samples is contained in the NW RFC Programming Guide, chapter 5.5 "Transactional RFC Server".

@bsrdjan : is this wrapped for Python?

Regards, Ulrich

**Comment by bsrdjan** [comment]
Hello @Ulrich-Schmidt,

the RfcInstallBgRfcHandlers() is not exposed yet. I defined 5 callback "C" functions in Python wrapper and did a quick test, it can be exposed.

Can the inner logic of callbacks be fixed (in "C"), or each callback shall be able to call Python method (like BAPI "User Exit") where user could add their own/custom logic?

**Comment by Ulrich-Schmidt** [comment]
@bsrdjan 
I would recommend to just call a Python method of the application in the C function, because the PyRFC framework cannot know, what the application wants to do. The application may want to do the transaction handling by a simple file-system based implementation (as in the sample code), or by using an Oracle database, or MS SQL Server, or MySQL, MongoDB or whatever...

If you do the implementation already in C, it would limit the application to one "fixed" approach. And I don't think the PyRFC framework would be the right layer for implementing this, even though it might be nice to relief the end-user from doing this low-level task...

**Comment by bsrdjan** [comment]
Hello @KunalBurangi,

RfcInstallBgRfcHandlers() RFC SDK method is now exposed in test branch [PyRFC/BgRfcHandlers](https://github.com/SAP/PyRFC/tree/BgRfcHandlers)

As per usage example in [bgrfc.py](https://github.com/SAP/PyRFC/blob/BgRfcHandlers/examples/server/bgrfc.py#L43), the  server instance is created as usual, and user-defined Python callback functions can be configured as follows

```python
def myCheckFunction(rfcHandle, unit_identifier):
    print("myCheckFunction", rfcHandle, unit_identifier)
    return 3

def myCommitFunction(rfcHandle, unit_identifier):
    print("myCommitFunction", rfcHandle, unit_identifier)
    return 2

server1 = Server({"dest": "gateway"}, {"dest": "MME"}, {"port": 8081, "server_log": False})

server1.bgrfc_init("SID", {"check": myCheckFunction, "commit": myCommitFunction})

server1.test()
```

"C" callback handlers are hard-coded in Python wrapper and they invoke Python handlers, with parameters: connection handle, unit identifier, unit state. The Server instance class (ie. server1) cannot be passed to Python handler because not known in "C" handler. It is not an issue I think but please check.

Would this interface be ok for your scenario?

**Comment by bsrdjan** [comment]
As this is new feature request, could you please create new issue and describe requirements there, so that we follow with open issue, not with this closed one.
